src/main/java/com/gregtechceu/gtceu/api/item/tool/TreeFellingHelper.java

```diff
@@ -52,23 +76,28 @@ public static void fellTree(ItemStack stack, Level level, BlockState origin, Blo
         }
 
         if (!visited.isEmpty() && miner instanceof ServerPlayer serverPlayer) {
-            List<BlockPos> orderedBlocks = visited.stream()
+            Deque<BlockPos> orderedBlocks = visited.stream()
                     .sorted(Comparator.comparingInt(pos -> pos.getY() - originPos.getY()))
                     .collect(Collectors.toCollection(LinkedList::new));
-            breakBlocksPerTick(serverPlayer, stack, orderedBlocks, origin.getBlock());
+            new TreeFellingHelper(serverPlayer, stack, orderedBlocks);
         }
     }
 
-    public static void breakBlocksPerTick(ServerPlayer player, ItemStack tool, List<BlockPos> posList,
-                                          Block originBlock) {
-        for (int i = 0; i < posList.size(); i++) {
-            int delayTick = i * 2; // 1 block per 2 tick
-            BlockPos pos = posList.get(i);
-            TaskHandler.enqueueServerTask(player.serverLevel(), () -> {
-                if (player.level().getBlockState(pos).is(originBlock)) {
-                    ToolHelper.breakBlockRoutine(player, tool, pos, true);
+    @SubscribeEvent
+    public static void onWorldTick(TickEvent.LevelTickEvent event) {
+        if (event.phase == TickEvent.Phase.START && event.side == LogicalSide.SERVER) {
+            for (var helper : helpers) {
+                if (event.level == helper.player.level()) {
+                    if (helper.orderedBlocks.isEmpty() || helper.tool.isEmpty() ||
+                            !getBehaviorsTag(helper.player.getMainHandItem()).getBoolean(TREE_FELLING_KEY)) {
```

Comment:
`getBehaviorsTag` will create the tag if it doesn't exist, so this can just add random tags to items.